packages/xsnap/test/test-xsnap.js

```diff
@@ -415,3 +415,123 @@ test('GC after snapshot vs restore', async t => {
   t.log({ beforeClone, workerGC, cloneGC, iters });
   t.is(workerGC, cloneGC);
 });
+
+(() => {
+  const grow = qty => `
+  const send = it => issueCommand(ArrayBuffer.fromString(JSON.stringify(it)));
+  let expr = \`"\${Array(${qty}).fill('abcd').join('')}"\`;
+  try {
+    eval(expr);
+    send(expr.length);
+  } catch (err) {
+    send(err.message);
+  }
+  `;
+  for (const debug of [false, true]) {
+    /**
+     * @type {Array<[number | undefined, number, string | null]>}
+     */
+    const tests = [
+      [undefined, 100, null],
+      [undefined, (8192 * 1024) / 4 + 100, 'buffer overflow'],
+      [2, 10, null],
+      [2, 50000, 'buffer overflow'],
+    ];
+    for (const [parserBufferSize, qty, failure] of tests) {
+      test(`parser buffer size ${parserBufferSize ||
+        'default'}k; rep ${qty}; debug ${debug}`, async t => {
+        const opts = { ...options(io), meteringLimit: 1e8, debug };
+        const vat = xsnap({ ...opts, parserBufferSize });
+        t.teardown(() => vat.terminate());
+        const expected = failure ? [failure] : [qty * 4 + 2];
+        // eslint-disable-next-line no-await-in-loop
+        await t.notThrowsAsync(vat.evaluate(grow(qty)));
+        t.deepEqual(
+          expected,
+          opts.messages.map(txt => JSON.parse(txt)),
+        );
+      });
+    }
+  }
+})();
+
+(() => {
+  const challenges = [
+    'new Uint8Array(2_130_706_417)',
+    'new Uint16Array(1_065_353_209)',
+    'new Uint32Array(532_676_605)',
+    'new BigUint64Array(266_338_303);',
+    'new Array(66_584_576).fill(0)',
+    '(new Array(66_584_575).fill(0))[66_584_575] = 0;',
+  ];
+
+  for (const statement of challenges) {
+    test(`large sizes - abort cluster: ${statement}`, async t => {
+      const opts = options(io);
+      const vat = xsnap(opts);
+      t.teardown(() => vat.terminate());
+      // eslint-disable-next-line no-await-in-loop
+      await t.throwsAsync(
+        vat.evaluate(`
+        (() => {
+          try {
+            // can't catch memory full
+            ${statement}\n
+          } catch (ignore) {
+            // ignore
+          }
+        })()`),
+        { code: ExitCode.E_NOT_ENOUGH_MEMORY },
+      );
+    });
+  }
+})();
+
+test('round-trip byte sequences from golang for dynamic IBC', async t => {
+  const opts = options(io);
+  const vat = xsnap(opts);
+  t.teardown(() => vat.terminate());
+
+  await vat.evaluate(`
+    globalThis.send = it => issueCommand(ArrayBuffer.fromString(JSON.stringify(it)));
```

Comment:
In order to call this a JS correctness test, let's use `TextEncoder` rather than the XS `ArrayBuffer.fromString` here.